### PR TITLE
add onLiftStart and onLiftEnd hooks

### PR DESCRIPTION
**Pull request by jaredcrowe**

### CODE REVIEW

**Comment by jaredcrowe** [comment]
@alexreardon just adding you to this so you're across it. I'll find someone from the AK team to review it in your absence

File: test/unit/integration/hooks-integration.spec.js [hunk 1]
```diff
@@ -26,6 +26,14 @@ const cancelWithKeyboard = dispatchWindowKeyDownEvent.bind(null, keyCodes.escape
 describe('hooks integration', () => {
   let hooks: Hooks;
   let wrapper;
+  // let callLog = [];
```
Comment (alexreardon) [inline test/unit/integration/hooks-integration.spec.js]: remove comments :)

File: src/state/reducer.js [hunk 2]
```diff
@@ -435,7 +447,13 @@ export default (state: State = clean('IDLE'), action: Action): State => {
     }
 
     const impact: DragImpact = result.impact;
-    const page: Position = result.pageCenter;
+    const currentMainAxisPosition: number =
```
Comment (alexreardon) [inline src/state/reducer.js]: should it have 'position' in the name when it is of type number?

File: src/state/action-creators.js [hunk 3]
```diff
@@ -457,26 +465,41 @@ export const lift = (id: DraggableId,
...
+      // Could improve this by explicitly waiting until all dimensions are published.
+      // Could also allow a lift to occur before all the dimensions are published
+      setTimeout(() => {
+        const newState: State = getState();
+
+        // drag was already cancelled before dimensions all collected
+        if (newState.phase !== 'COLLECTING_DIMENSIONS') {
+          return;
+        }
+        const realClient = {
```
Comment (alexreardon) [inline src/state/action-creators.js]: can you explain this one?

File: src/state/action-creators.js [hunk 4]
```diff
@@ -457,26 +465,41 @@ export const lift = (id: DraggableId,
...
-      const newState: State = getState();
-
-      // drag was already cancelled before dimensions all collected
-      if (newState.phase !== 'COLLECTING_DIMENSIONS') {
-        return;
-      }
-      dispatch(completeLift(id, type, client, windowScroll, isScrollAllowed));
+      dispatch(requestDimensions(type));
+
+      // Dimensions will be requested synchronously
```
Comment (alexreardon) [inline src/state/action-creators.js]: can you please reword this - i am finding it hard to follow

File: src/state/action-creators.js [hunk 5]
```diff
@@ -457,26 +465,41 @@ export const lift = (id: DraggableId,
...
+          return;
+        }
+        const realClient = {
+          center: newState.dimension.draggable[id].client.withMargin.center,
+          selection: client.selection,
+        };
+
+        dispatch(completeLift());
+
+        // Wait for a re-render again
```
Comment (alexreardon) [inline src/state/action-creators.js]: can you please explain why this wait it needed

File: src/state/action-creators.js [hunk 6]
```diff
@@ -457,26 +465,41 @@ export const lift = (id: DraggableId,
...
-        return;
-      }
-      dispatch(completeLift(id, type, client, windowScroll, isScrollAllowed));
+      dispatch(requestDimensions(type));
+
+      // Dimensions will be requested synchronously
+      // after they are done - lift.
+      // Could improve this by explicitly waiting until all dimensions are published.
+      // Could also allow a lift to occur before all the dimensions are published
+      setTimeout(() => {
```
Comment (alexreardon) [inline src/state/action-creators.js]: what happens if a cancel happens during these phases? Eg a quick release

**Comment by alexreardon** [comment]
Can you please add an explanation about the rationale for this change and the approach taken? Also, I am worried that these chained set timeouts are susceptible to quick cancels - how do we prevent against this?
